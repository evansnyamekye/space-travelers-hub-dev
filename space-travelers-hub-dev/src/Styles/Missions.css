.mission-page {
  display: flex;
  align-items: center;
  justify-content: center;
}

.missionTable {
  width: 95%;
  border-collapse: collapse;
  text-align: left;
}

.missionTable,
td,
th {
  border: 1px solid;
}

.missionTable td,
.missionTable th {
  border: 1px solid #ddd;
  padding: 0.5rem;
}

.mission-name {
  vertical-align: top;
}

.description {
  padding-bottom: 1rem;
}

.mission-btn {
  text-align: center;
}

.missionTable tr:nth-child(odd) { background-color: #f2f2f2; }

.active-member {
  background-color: #2ea8d1;
  color: white;
  text-align: center;
  border-radius: 0.2rem;
  padding: 0.2rem;
}

.not-member {
  background-color: gray;
  color: white;
  text-align: center;
  border-radius: 0.2rem;
  padding: 0.2rem;
}

.leave-mission {
  width: 8rem;
  padding: 0.5rem;
  background-color: transparent;
  border: 1px solid red;
  color: red;
  border-radius: 0.3rem;
  cursor: pointer;
}

.join-mission {
  width: 8rem;
  padding: 0.5rem;
  border-radius: 0.3rem;
  cursor: pointer;
}
