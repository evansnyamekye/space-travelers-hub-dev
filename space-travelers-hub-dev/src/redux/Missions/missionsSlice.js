import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const apiLink = 'https://api.spacexdata.com/v3/missions';

export const getMissionsData = createAsyncThunk('missions/getMissionsData', async (thunkAPI) => {
  try {
    const response = await axios.get(apiLink);
    return response.data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.message);
  }
});

const missionsSlice = createSlice({
  name: 'missions',
  initialState: {
    missions: [],
    isLoading: false,
    error: null,
  },
  reducers: {
    joinMission: (state, action) => {
      const id = action.payload;
      state.missions = state.missions.map((mission) => (
        {
          ...mission,
          mission_join: (mission.mission_id
          === id) ? true : mission.mission_join,
        }
      ));
    },

    cancelMission: (state, action) => {
      state.missions = state.missions.map((mission) => (
        {
          ...mission,
          mission_join: action.payload
          === mission.mission_id ? false : mission.mission_join,
        }
      ));
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMissionsData.pending, (state) => {
        state.isLoading = false;
      })
      .addCase(getMissionsData.fulfilled, (state, action) => {
        state.isLoading = false;
        state.missions = action.payload.map((mission) => ({
          mission_id: mission.mission_id,
          mission_name: mission.mission_name,
          description: mission.description,
          mission_join: false,
        }));
      })
      .addCase(getMissionsData.rejected, (state, action) => {
        state.isLoading = false;
        state.error = action.error;
      });
  },
});

export const { joinMission, cancelMission } = missionsSlice.actions;
export default missionsSlice.reducer;
