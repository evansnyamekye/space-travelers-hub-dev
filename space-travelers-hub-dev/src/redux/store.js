import { configureStore } from '@reduxjs/toolkit';
import rocketReduser from './rocketSlice';
import missionReducer from './Missions/missionsSlice';

const store = configureStore({
  reducer: {
    rockets: rocketReduser,
    missions: missionReducer,
  },
});

export default store;
