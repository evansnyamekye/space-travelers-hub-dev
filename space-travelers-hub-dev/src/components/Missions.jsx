import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getMissionsData } from '../redux/Missions/missionsSlice';
import Mission from './Mission';
import '../Styles/Missions.css';

const Missions = () => {
  const { isLoading, error, missions } = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getMissionsData());
  }, []);
  if (isLoading) {
    return <div>loading...</div>;
  }
  if (error) {
    return <div style={{ color: 'red' }}>There is an error...!</div>;
  }

  return (
    <div className="mission-page">
      <table className="missionTable">
        <thead className="missionThead">
          <th width="10%">
            <h3>Mision</h3>
          </th>

          <th width="65%">
            <h3>Description</h3>
          </th>

          <th width="12.5%">
            <h3>Status</h3>
          </th>
          <th width="12.5%">
            {' '}
          </th>
        </thead>

        <tbody>
          {

          missions.map((mission) => (
            <Mission
              key={mission.mission_id}
              mission_id={mission.mission_id}
              mission_name={mission.mission_name}
              description={mission.description}
              mission_join={mission.mission_join}
            />
          ))

        }

        </tbody>
      </table>
    </div>
  );
};
export default Missions;
