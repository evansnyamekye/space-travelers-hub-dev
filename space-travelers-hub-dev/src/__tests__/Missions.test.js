import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import Missions from '../components/Missions';
import store from '../redux/store';

describe('Test Missions components', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
