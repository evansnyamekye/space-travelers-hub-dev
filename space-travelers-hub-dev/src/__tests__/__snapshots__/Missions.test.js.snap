// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Missions components renders correctly 1`] = `
<div>
  <div
    class="mission-page"
  >
    <table
      class="missionTable"
    >
      <thead
        class="missionThead"
      >
        <th
          width="10%"
        >
          <h3>
            Mision
          </h3>
        </th>
        <th
          width="65%"
        >
          <h3>
            Description
          </h3>
        </th>
        <th
          width="12.5%"
        >
          <h3>
            Status
          </h3>
        </th>
        <th
          width="12.5%"
        >
           
        </th>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
